### PR TITLE
Torchmetrics paper

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Torchmetrics paper candidate for submission in https://joss.theoj.org/

The idea is to write a high-level summary/paper about TM and publish it with JOSS. JOSS format is very short, and the end goal is to have a reference for the repository that users can refer to in their own publications. If TM gets more traction and reaches v1.0, we can reevaluate if a longer paper should be written explaining in more detail the core design.

@lucadiliello, @stancld, @quancs we want to thank you for contributing the main part of a domain (information retrieval, NLP, audio) by offering you an co-authorship on the paper. If you do not want this, please tell us and we will remove you from the paper. If you want this, please read through the paper and if you agree to the content please suggest changes on how you want to be credited (name + affiliation).

@teddykoker, @ananyahjha93 you were both essential to developing the core design of torchmetrics. You are of cause both offered co-authorship. Please read through the paper and make any comment + suggestion you want (also if we should change affiliation for you).

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
In short, we expect that all listed authors shall approve this PR as a sign that they agree with their authorship and text content
Then the next step after merging will be sending a submission to the JOSS publication :rabbit:

File: docs/paper/paper.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
+---
+title: TorchMetrics - Measuring Reproducibility in PyTorch
+tags:
+  - python
+  - deep learning
+  - pytorch
+authors:
+  - Nicki Skafte Detlefsen
+    affiliation: "1,2"
+  - Jirka Borovec
```
Comment (Borda) [inline docs/paper/paper.md:10]:
```suggestion
  - Jiri Borovec
```

File: docs/paper/paper.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    affiliation: "1"
+  - Justus Schock
+    affiliation: "1"
+  - Ananya Harsh Jha
+    affiliation: "1"
+  - Teddy Koker
+    affiliation: "1"
+  - Lucadiliello
```
Comment (Borda) [inline docs/paper/paper.md:18]: @lucadiliello ^^

File: docs/paper/paper.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  - Stancld
```
Comment (Borda) [inline docs/paper/paper.md:19]: @Stancld ^^

File: docs/paper/paper.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  - Quancs
```
Comment (Borda) [inline docs/paper/paper.md:20]: @quancs ^^

File: docs/paper/paper.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        self.add_state("correct", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("total", default=torch.tensor(0), dist_reduce_fx="sum")
+
+    def update(self, preds: torch.Tensor, target: torch.Tensor) -> None:
+        # update takes preds and target and accumulate the current stream
+        # of data into the global states for later
+        self.correct += torch.sum(preds == target)
+        self.total += target.numel()
+
+    def compute(self) -> Tensor:
```
Comment (Borda) [inline docs/paper/paper.md:71]:
```suggestion
    def compute(self) -> torch.Tensor:
```

File: docs/paper/paper.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+TorchMetrics exhibits high test coverage on the various configurations, including all three major OS platforms (Linux, macOS and Windows), Python, CUDA and PyTorch versions. We test both minimum and latest package requirements for all combinations of OS and Python versions and include additional tests for each PyTorch version from 1.3 up to future development versions. As the development cycle, we run a full test suite for each merge to master branch and a significant portion of these tests for each open PR by any contributor. All standard tests are running on CPU, but as Deep Learning is mainly focused on GPU computing, we are running all tests also in a multi-GPU setting in the cloud. For usability, we have auto-generated HTML documentation at https://torchmetrics.readthedocs.io/en/latest/ from the source code to keep it tied with all API changes.
```
Comment (Borda) [inline docs/paper/paper.md:80]:
```suggestion
TorchMetrics exhibits high test coverage on the various configurations, including all three major OS platforms (Linux, macOS and Windows), Python, CUDA and PyTorch versions. We test both minimum and latest package requirements for all combinations of OS and Python versions and include additional tests for each PyTorch version from 1.3 up to future development versions. As the development cycle, we run a full test suite for each merge to master branch and a significant portion of these tests for each open PR by any contributor. All standard tests are running on CPU, but as Deep Learning is mainly focused on GPU computing, we are running all tests also in a multi-GPU setting in the cloud. For usability, we have auto-generated HTML documentation at https://torchmetrics.readthedocs.io/en/stable/ from the source code to keep it tied with all API changes.
```

**Comment by teddykoker** [comment]
Great idea! Looks good to me :)

**Review by teddykoker — APPROVED** [review]

File: docs/paper/paper.md (hunk 2 above)
Comment (lucadiliello) [reply docs/paper/paper.md:18]: My name is "Luca Di Liello" and my affiliation "University of Trento". Thank you a lot for letting us be part of this paper :)

Comment (Borda) [reply docs/paper/paper.md]: mind rereview and eventually approve? :]

File: docs/paper/paper.md (hunk 3 above)
Comment (stancld) [reply docs/paper/paper.md:20]:
```suggestion
  - Daniel Stancl
```
affiliation: Charles University

**Comment by stancld** [comment]
I think this is a great idea! Thank you very much for involving me, I'm honoured O:]

File: docs/paper/paper.bib [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    primaryClass={cs.CY}
+}
+@misc{fid,
+    title={GANs Trained by a Two Time-Scale Update Rule Converge to a Local Nash Equilibrium},
+    author={Martin Heusel and Hubert Ramsauer and Thomas Unterthiner and Bernhard Nessler and Sepp Hochreiter},
+    year={2018},
+    eprint={1706.08500},
+    archivePrefix={arXiv},
+    primaryClass={cs.LG}
+}
```
Comment (stancld) [inline docs/paper/paper.bib:16]: The paper was also presented at NeurIPS 2017

```
@inproceedings{fid,
    title = {{GANs} {Trained} by a {Two} {Time}-{Scale} {Update} {Rule} {Converge} to a {Local} {Nash} {Equilibrium}},
    volume = {30},
    url = {https://proceedings.neurips.cc/paper/2017/hash/8a1d694707eb0fefe65871369074926d-Abstract.html},
    booktitle = {Advances in {Neural} {Information} {Processing} {Systems}},
    publisher = {Curran Associates, Inc.},
    author = {Heusel, Martin and Ramsauer, Hubert and Unterthiner, Thomas and Nessler, Bernhard and Hochreiter, Sepp},
    pages={6629--6640},
    year = {2017},
}
```

File: docs/paper/paper.md (hunk 3 above)
Comment (Borda) [reply docs/paper/paper.md]: mind rereview after the update? :rabbit:

**Review by stancld — APPROVED** [review]

File: docs/paper/paper.md (hunk 2 above)
Comment (lucadiliello) [reply docs/paper/paper.md]: I think I can not.

Comment (Borda) [reply docs/paper/paper.md]: I think you can, just it wont be green :heavy_check_mark:  but gray :white_check_mark: 
image

File: docs/paper/paper.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  - python
+  - deep learning
+  - pytorch
+authors:
+  - Nicki Skafte Detlefsen
+    affiliation: "1,2"
+  - Jiri Borovec
+    affiliation: "1"
+  - Justus Schock
+    affiliation: "1"
```
Comment (justusschock) [inline docs/paper/paper.md:13]:
```suggestion
    affiliation: "1,3"
```

File: docs/paper/paper.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+  - Ananya Harsh Jha
+    affiliation: "1"
+  - Teddy Koker
+    affiliation: "1"
+  - Luca Di Liello
+    affiliation: "3"
```
Comment (justusschock) [inline docs/paper/paper.md:19]:
```suggestion
    affiliation: "4"
```

File: docs/paper/paper.md [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+affiliations:
+ - name: Grid AI
+   index: 1
+ - name: Technical University of Denmark
+   index: 2
+ - name: University of Trento
+   index: 3
+ - name: Charles University
+   index: 4
```
Comment (justusschock) [inline docs/paper/paper.md:34]:
```suggestion
 - name: University Hospital Düsseldorf
   index: 3
 - name: University of Trento
   index: 4
 - name: Charles University
   index: 5
```

File: docs/paper/paper.md [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+  - Justus Schock
+    affiliation: "1"
+  - Ananya Harsh Jha
+    affiliation: "1"
+  - Teddy Koker
+    affiliation: "1"
+  - Luca Di Liello
+    affiliation: "3"
+  - Daniel Stancl
+    affiliation: "4"
```
Comment (justusschock) [inline docs/paper/paper.md:21]:
```suggestion
    affiliation: "5"
```

**Review by justusschock — APPROVED** [review]

**Review by quancs — APPROVED** [review]

File: docs/paper/paper.md (hunk 4 above)
Comment (quancs) [reply docs/paper/paper.md:22]:
```suggestion
  - Changsheng Quan
    affiliation: "6"
```

File: docs/paper/paper.md [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+ - name: Grid AI
+   index: 1
+ - name: Technical University of Denmark
+   index: 2
+ - name: University Hospital Düsseldorf
+   index: 3
+ - name: University of Trento
+   index: 4
+ - name: Charles University
+   index: 5
```
Comment (quancs) [inline docs/paper/paper.md:36]:
```suggestion
   index: 5
 - name: Zhejiang University
   index: 6
```

**Review by quancs — APPROVED** [review]

File: docs/paper/paper.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+ - name: Zhejiang University
+   index: 6
```
Comment (Borda) [inline docs/paper/paper.md:39]:
```suggestion
   index: 6
 - name: New York University
   index: 7
```

File: docs/paper/paper.md [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+  - Teddy Koker
+    affiliation: "1"
+  - Luca Di Liello
+    affiliation: "4"
+  - Daniel Stancl
+    affiliation: "5"
+  - Changsheng Quan
+    affiliation: "6"
+  - William Falcon
+    affiliation: "1"
```
Comment (Borda) [inline docs/paper/paper.md:25]:
```suggestion
    affiliation: "1,7"
```

File: docs/paper/paper.bib [hunk 15]
```diff
@@ -0,0 +1,81 @@
+@misc{reproduceability,
```
Comment (justusschock) [inline docs/paper/paper.bib:1]:
```suggestion
@misc{reproducibility,
```

File: docs/paper/paper.md [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    index: 5
+  - name: Zhejiang University
+    index: 6
+date: 08 Dec 2021
+bibliography: paper.bib
+---
+
+# Summary
+
+Among recent machine learning publications, many cannot be reproduced \[@reproducibility\]. This is often due to different metrics implementations, which may lead to important training mechanisms like checkpointing, learning rate schedules, or early stopping to behave differently. For example, complex metrics like Fréchet inception distance (FID) for synthetic image quality evaluation \[@fid\] may differ widely based on the specific interpolation method used.
```
Comment (justusschock) [inline docs/paper/paper.md:46]:
```suggestion
Among recent machine learning publications, many cannot be reproduced [@reproducibility]. This is often due to different metrics implementations, which may lead to important training mechanisms like checkpointing, learning rate schedules, or early stopping to behave differently. For example, complex metrics like Fréchet inception distance (FID) for synthetic image quality evaluation \[@fid\] may differ widely based on the specific interpolation method used.
```

File: docs/paper/paper.md [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+A metric\[^1\] hereby is an objective quantitative measure of the model’s performance. To tackle the problem of reproducibility, there are initiatives such as Papers With Code \[@papers_with_code\], which link the research code with the corresponding papers. However, not all authors provide their code openly accessible. Therefore, we aim to tackle the problem another way - We believe that unified implementations of the relevant metrics can help overcome reproducibility by providing a de-facto reference implementation.
```
Comment (justusschock) [inline docs/paper/paper.md:48]:
```suggestion
A metric[^1] hereby is an objective quantitative measure of the model’s performance. To tackle the problem of reproducibility, there are initiatives such as Papers With Code [@papers_with_code], which link the research code with the corresponding papers. However, not all authors provide their code openly accessible. Therefore, we aim to tackle the problem another way - We believe that unified implementations of the relevant metrics can help overcome reproducibility by providing a de-facto reference implementation.
```

File: docs/paper/paper.md [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+
+\[^1\]: Not to be confused with mathematical definitions of metrics as per https://en.wikipedia.org/wiki/Metric_(mathematics) whose properties do not have to hold for the machine learning understanding of the same word.
```
Comment (justusschock) [inline docs/paper/paper.md:50]:
```suggestion
[^1]: Not to be confused with mathematical definitions of metrics as per https://en.wikipedia.org/wiki/Metric_(mathematics) whose properties do not have to hold for the machine learning understanding of the same word.
```

File: docs/paper/paper.md [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+
+TorchMetrics is a general-purpose metrics package covering a wide variety of tasks and domains. It includes standard classification and regression metrics and some domain-specific, for example, audio, computer vision, natural language processing, and information retrieval. In the case of particular metrics, we integrate with a well-tested and established third-party package as a starter, and later on, we re-implement them as PyTorch \[@pytorch\] native to keep minimal dependencies, make use of available hardware accelerators (mainly GPUs), avoid inter-device transfers whenever possible and maintain full control over each metric implementation.
```
Comment (justusschock) [inline docs/paper/paper.md:52]:
```suggestion
TorchMetrics is a general-purpose metrics package covering a wide variety of tasks and domains. It includes standard classification and regression metrics and some domain-specific, for example, audio, computer vision, natural language processing, and information retrieval. In the case of particular metrics, we integrate with a well-tested and established third-party package as a starter, and later on, we re-implement them as PyTorch [@pytorch] native to keep minimal dependencies, make use of available hardware accelerators (mainly GPUs), avoid inter-device transfers whenever possible and maintain full control over each metric implementation.
```

File: docs/paper/paper.md [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+
+# Statement of need
+
+Currently, there exists no native package in PyTorch that implements commonly-used metrics within machine learning that is commonly adopted. Some native PyTorch libraries support domain-specific metrics such as Transformers \[@transformers\] for calculating NLP-specific metrics. However, no library exists that covers multiple domains. PyTorch users, therefore, often rely on non-PyTorch packages such as Scikit-learn \[@scikit_learn\] for computing even simple metrics such as accuracy, F1, or AUROC metrics.
```
Comment (justusschock) [inline docs/paper/paper.md:56]:
```suggestion
Currently, there exists no native package in PyTorch that implements commonly-used metrics within machine learning that is commonly adopted. Some native PyTorch libraries support domain-specific metrics such as Transformers [@transformers] for calculating NLP-specific metrics. However, no library exists that covers multiple domains. PyTorch users, therefore, often rely on non-PyTorch packages such as Scikit-learn [@scikit_learn] for computing even simple metrics such as accuracy, F1, or AUROC metrics.
```

File: docs/paper/paper.md [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+
+TorchMetrics solves this problem by introducing stateful metrics that can calculate metric values on a stream of data alongside the classical functional and stateless metrics provided by other packages like Scikit-learn. We do this with an effortless `update` and `compute` interface, well known from packages such as Keras \[@keras\]. The `update` function takes in a batch of predictions and targets and updates the internal state. For example, for a metric such as accuracy, the internal states are simply the number of correctly classified samples and the total observed number of samples. When all batches have been passed to the `update` method, the `compute` method can get the accumulated accuracy over all the batches. In addition to `update` and `compute`, each metric also has a `forward` method (as any other `torch.nn.Module`) that can be used to both get the metric on the current batch of data and accumulate global state. This enables the user to get fine-grained info about the metric on the individual batch and the global metric of how well their model is doing.
```
Comment (justusschock) [inline docs/paper/paper.md:60]:
```suggestion
TorchMetrics solves this problem by introducing stateful metrics that can calculate metric values on a stream of data alongside the classical functional and stateless metrics provided by other packages like Scikit-learn. We do this with an effortless `update` and `compute` interface, well known from packages such as Keras [@keras]. The `update` function takes in a batch of predictions and targets and updates the internal state. For example, for a metric such as accuracy, the internal states are simply the number of correctly classified samples and the total observed number of samples. When all batches have been passed to the `update` method, the `compute` method can get the accumulated accuracy over all the batches. In addition to `update` and `compute`, each metric also has a `forward` method (as any other `torch.nn.Module`) that can be used to both get the metric on the current batch of data and accumulate global state. This enables the user to get fine-grained info about the metric on the individual batch and the global metric of how well their model is doing.
```

File: docs/paper/paper.md [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+        # of data into the global states for later
+        self.correct += torch.sum(preds == target)
+        self.total += target.numel()
+
+    def compute(self) -> torch.Tensor:
+        # compute takes the accumulated states and returns the final metric value
+        return self.correct / self.total
+```
+
+Another core feature of TorchMetrics is its ability to scale to multiple devices seamlessly. Modern deep learning models are often trained on hundreds of devices GPUs or TPUs (see \[@large_example1\]\[@large_example2\] for examples), and the corresponding metrics calculated during training and evaluation, therefore, need to be synchronized to get the correct value. TorchMetrics completely takes care of this in the background, automatically detecting if a metric is being updated on multiple devices and accumulating the states from different devices before reporting the calculated metric to the user.
```
Comment (justusschock) [inline docs/paper/paper.md:87]:
```suggestion
Another core feature of TorchMetrics is its ability to scale to multiple devices seamlessly. Modern deep learning models are often trained on hundreds of devices GPUs or TPUs (see [@large_example1][@large_example2] for examples), and the corresponding metrics calculated during training and evaluation, therefore, need to be synchronized to get the correct value. TorchMetrics completely takes care of this in the background, automatically detecting if a metric is being updated on multiple devices and accumulating the states from different devices before reporting the calculated metric to the user.
```

File: docs/paper/paper.md [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+  - name: Luca Di Liello
+    affiliation: '4'
+  - name: Daniel Stancl
+    affiliation: '5'
+  - name: Changsheng Quan
+    affiliation: '6'
+  - name: William Falcon
+    affiliation: '1'
+affiliations:
+  - name: Grid AI
```
Comment (Borda) [inline docs/paper/paper.md:28]:
```suggestion
  - name: Grid AI labs
```

File: docs/paper/paper.md [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+        # of data into the global states for later
+        self.correct += torch.sum(preds == target)
+        self.total += target.numel()
+
+    def compute(self) -> torch.Tensor:
+        # compute takes the accumulated states and returns the final metric value
+        return self.correct / self.total
+```
+
+Another core feature of TorchMetrics is its ability to scale to multiple devices seamlessly. Modern deep learning models are often trained on hundreds of devices GPUs or TPUs (see [@large_example1][@large_example2] for examples), and the corresponding metrics calculated during training and evaluation, therefore, need to be synchronized to get the correct value. TorchMetrics completely takes care of this in the background, automatically detecting if a metric is being updated on multiple devices and accumulating the states from different devices before reporting the calculated metric to the user.
```
Comment (justusschock) [inline docs/paper/paper.md:87]:
```suggestion
Another core feature of TorchMetrics is its ability to scale to multiple devices seamlessly. Modern deep learning models are often trained on hundreds of devices GPUs or TPUs (see [@large_example1; @large_example2] for examples), and the corresponding metrics calculated during training and evaluation, therefore, need to be synchronized to get the correct value. TorchMetrics completely takes care of this in the background, automatically detecting if a metric is being updated on multiple devices and accumulating the states from different devices before reporting the calculated metric to the user.
```

File: docs/paper/paper.md [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+  - name: Luca Di Liello
+    affiliation: '4'
+  - name: Daniel Stancl
+    affiliation: '5'
+  - name: Changsheng Quan
+    affiliation: '6'
+  - name: William Falcon
+    affiliation: '1'
+affiliations:
+  - name: Grid AI labs
```
Comment (justusschock) [inline docs/paper/paper.md:28]:
```suggestion
  - name: Grid AI Labs
```

@Borda should this be capitalized?

File: docs/paper/paper.md [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+  - name: Teddy Koker
+    affiliation: '1'
+  - name: Luca Di Liello
+    affiliation: '4'
+  - name: Daniel Stancl
+    affiliation: '5'
+  - name: Changsheng Quan
+    affiliation: '6'
+  - name: William Falcon
+    affiliation: '1'
```
Comment (Borda) [inline docs/paper/paper.md:26]:
```suggestion
    affiliation: '1,7'
```

File: docs/paper/paper.md [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+  - name: Technical University of Denmark
+    index: 2
+  - name: University Hospital Düsseldorf
+    index: 3
+  - name: University of Trento
+    index: 4
+  - name: Charles University
+    index: 5
+  - name: Zhejiang University
+    index: 6
```
Comment (Borda) [inline docs/paper/paper.md:39]:
```suggestion
    index: 6
  - name: New York University
    index: 7
```

File: docs/paper/paper.md [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+
+TorchMetrics is released under the Apache 2.0 license, and the source code can be found at https://github.com/PytorchLightning/metrics.
+
+# Acknowledgement
+
+The TorchMetrics team thanks Thomas Chaton, Ethan Harris, Carlos Mocholí, Sean Naren and Adrian Wälchli for contributing ideas, discussions on API design, and contributing with Pull Request reviews. We also thank all our open-source contributors for reporting and resolving issues with this package. We are grateful to the PyTorch Lightning team for their ongoing and dedicated support of this project, and Grid.ai for providing computing resources and cloud credits needed to run our Continues Integrations.
```
Comment (justusschock) [inline docs/paper/paper.md:114]:
```suggestion
The TorchMetrics team thanks Thomas Chaton, Ethan Harris, Carlos Mocholí, Sean Narenthiran and Adrian Wälchli for contributing ideas, discussions on API design, and contributing with Pull Request reviews. We also thank all our open-source contributors for reporting and resolving issues with this package. We are grateful to the PyTorch Lightning team for their ongoing and dedicated support of this project, and Grid.ai for providing computing resources and cloud credits needed to run our Continues Integrations.
```

**Review by lucadiliello — APPROVED** [review]

File: docs/paper/paper.md (hunk 2 above)
Comment (lucadiliello) [reply docs/paper/paper.md]: Ohhh didn't know about that option, done!

File: docs/paper/paper.md [hunk 29]
```diff
@@ -0,0 +1,104 @@
...
+  - name: Teddy Koker
+    affiliation: '1'
+  - name: Luca Di Liello
+    affiliation: '4'
+  - name: Daniel Stancl
+    affiliation: '5'
+  - name: Changsheng Quan
+    affiliation: '6'
+  - name: William Falcon
+    affiliation: '1,7'
```
Comment (stancld) [inline docs/paper/paper.md:26]: In the case of multiple affiliations, should it be unified? I.e. either '1,2' or 1,2? (See lines 9, 13 and 26)

Comment (justusschock) [reply docs/paper/paper.md:26]: They need to be with quotes as stated by the template :)

File: docs/paper/paper.md [hunk 30]
```diff
@@ -0,0 +1,104 @@
+---
+title: TorchMetrics - Measuring Reproducibility in PyTorch
+tags:
+  - python
+  - deep learning
+  - pytorch
+authors:
+  - name: Nicki Skafte Detlefsen
+    affiliation: 1,2
```
Comment (justusschock) [inline docs/paper/paper.md:9]:
```suggestion
    affiliation: '1,2'
```

File: docs/paper/paper.md [hunk 31]
```diff
@@ -0,0 +1,104 @@
...
+  - name: Jiri Borovec
+    affiliation: '1'
+  - name: Justus Schock
+    affiliation: 1,3
```
Comment (justusschock) [inline docs/paper/paper.md:13]:
```suggestion
    affiliation: '1,3'
```

File: docs/paper/paper.md (hunk 29 above)
Comment (justusschock) [reply docs/paper/paper.md]: fixed :)

**Review by ananyahjha93 — APPROVED** [review]

**Comment by Borda** [comment]
@williamFalcon could you please review and approve this publication :rabbit:

File: docs/paper/paper.md [hunk 32]
```diff
@@ -0,0 +1,106 @@
...
+
+TorchMetrics is released under the Apache 2.0 license, and the source code can be found at https://github.com/PytorchLightning/metrics.
+
+# Acknowledgement
+
+The TorchMetrics team thanks Thomas Chaton, Ethan Harris, Carlos Mocholí, Sean Narenthiran and Adrian Wälchli for contributing ideas, discussions on API design, and contributing with Pull Request reviews. We also thank all our open-source contributors for reporting and resolving issues with this package. We are grateful to the PyTorch Lightning team for their ongoing and dedicated support of this project, and Grid.ai for providing computing resources and cloud credits needed to run our Continues Integrations.
```
Comment (Borda) [inline docs/paper/paper.md:104]:
```suggestion
The TorchMetrics team thanks Thomas Chaton, Ethan Harris, Carlos Mocholí, Sean Narenthiran, and Adrian Wälchli for contributing ideas, participating in discussions on API design, and completing Pull Request reviews. We also thank all of our open-source contributors for reporting and resolving issues with this package. We are grateful to the PyTorch Lightning team for their ongoing and dedicated support of this project, and Grid.ai for providing computing resources and cloud credits needed to run our Continuos Integrations.
```

File: docs/paper/paper.md [hunk 33]
```diff
@@ -0,0 +1,106 @@
...
+
+TorchMetrics is released under the Apache 2.0 license, and the source code can be found at https://github.com/PytorchLightning/metrics.
+
+# Acknowledgement
+
+The TorchMetrics team thanks Thomas Chaton, Ethan Harris, Carlos Mocholí, Sean Narenthiran, and Adrian Wälchli for contributing ideas, participating in discussions on API design, and completing Pull Request reviews. We also thank all of our open-source contributors for reporting and resolving issues with this package. We are grateful to the PyTorch Lightning team for their ongoing and dedicated support of this project, and Grid.ai for providing computing resources and cloud credits needed to run our Continuos Integrations.
```
Comment (Borda) [inline docs/paper/paper.md:104]:
```suggestion
The TorchMetrics team thanks Thomas Chaton, Ethan Harris, Carlos Mocholí, Sean Narenthiran, Adrian Wälchli, Maxim Grechkin, and Ananth Subramaniam for contributing ideas, participating in discussions on API design, and completing Pull Request reviews. We also thank all of our open-source contributors for reporting and resolving issues with this package. We are grateful to the PyTorch Lightning team for their ongoing and dedicated support of this project, and Grid.ai for providing computing resources and cloud credits needed to run our Continuos Integrations.
```

**Comment by Borda** [comment]
@lantiga could you pls have the final read and approve hopefully... :rabbit:

**Review by lantiga — APPROVED** [review]
LGTM!

File: docs/paper/paper.md [hunk 34]
```diff
@@ -0,0 +1,106 @@
...
+
+TorchMetrics exhibits high test coverage on the various configurations, including all three major OS platforms (Linux, macOS and Windows), Python, CUDA and PyTorch versions. We test both minimum and latest package requirements for all combinations of OS and Python versions and include additional tests for each PyTorch version from 1.3 up to future development versions. As the development cycle, we run a full test suite for each merge to master branch and a significant portion of these tests for each open PR by any contributor. All standard tests are running on CPU, but as Deep Learning is mainly focused on GPU computing, we are running all tests also in a multi-GPU setting in the cloud. For usability, we have auto-generated HTML documentation at https://torchmetrics.readthedocs.io/en/stable/ from the source code to keep it tied with all API changes.
```
Comment (Borda) [inline docs/paper/paper.md:98]:
```suggestion
TorchMetrics exhibits high test coverage on the various configurations, including all three major OS platforms (Linux, macOS, and Windows), and various Python, CUDA, and PyTorch versions. We test both minimum and latest package requirements for all combinations of OS and Python versions and include additional tests for each PyTorch version from 1.3 up to future development versions. As the development cycle, we run a full test suite for each merge to master branch and a significant portion of these tests for each open PR by any contributor. All standard tests are running on CPU, but as Deep Learning is mainly focused on GPU computing, we run all tests also in a multi-GPU setting. For usability, we have auto-generated HTML documentation (hosted at https://torchmetrics.readthedocs.io/en/stable/) from the source code to keep it tied with all API changes.
```

File: docs/paper_JOSS/paper.md [hunk 35]
```diff
@@ -0,0 +1,106 @@
...
+    index: 6
+  - name: New York University
+    index: 7
+date: 08 Dec 2021
+bibliography: paper.bib
+---
+
+# Summary
+
+Among recent machine learning publications, many cannot be reproduced [@reproducibility]. This is often due to different metrics implementations, which may lead to important training mechanisms like checkpointing, learning rate schedules, or early stopping to behave differently. For example, complex metrics like Fréchet inception distance (FID) for synthetic image quality evaluation [@fid] may differ widely based on the specific interpolation method used.
```
Comment (williamFalcon) [inline docs/paper_JOSS/paper.md:50]: check grammar...

metrics implementations -> metric implementations

File: docs/paper_JOSS/paper.md [hunk 36]
```diff
@@ -0,0 +1,106 @@
...
+
+A metric[^1] hereby is an objective quantitative measure of the model’s performance. To tackle the problem of reproducibility, there are initiatives such as Papers With Code [@papers_with_code], which link the research code with the corresponding papers. However, not all authors provide their code openly accessible. Therefore, we aim to tackle the problem another way - We believe that unified implementations of the relevant metrics can help overcome reproducibility by providing a de-facto reference implementation.
```
Comment (williamFalcon) [inline docs/paper_JOSS/paper.md:52]: hereby is unnecessary (don't add unnecessary language)

Comment (williamFalcon) [reply docs/paper_JOSS/paper.md:52]: not all authors provide their code openly accessible -> not all authors make opensource their code

Comment (williamFalcon) [reply docs/paper_JOSS/paper.md:52]: remove (herefore, we aim to tackle the problem another way -) (unecessary)..

File: docs/paper_JOSS/paper.md [hunk 37]
```diff
@@ -0,0 +1,104 @@
...
+        self.correct += torch.sum(preds == target)
+        self.total += target.numel()
+
+    def compute(self) -> Tensor:
+        # compute takes the accumulated states
+        # and returns the final metric value
+        return self.correct / self.total
+```
+
+Another core feature of TorchMetrics is its ability to scale to multiple devices seamlessly. Modern deep learning models are often trained on hundreds of devices GPUs or TPUs (see [@large_example1; @large_example2] for examples), and the corresponding metrics calculated during training and evaluation, therefore, need to be synchronized to get the correct value. TorchMetrics completely takes care of this in the background, automatically detecting if a metric is being updated on multiple devices and accumulating the states from different devices before reporting the calculated metric to the user.
```
Comment (williamFalcon) [inline docs/paper_JOSS/paper.md:92]:
```suggestion
Another core feature of TorchMetrics is its ability to scale to multiple devices seamlessly. Modern deep learning models are often trained on hundreds of devices GPUs or TPUs (see [@large_example1; @large_example2] for examples), and the corresponding metrics calculated during training and evaluation, therefore, need to be synchronized to get the correct value. TorchMetrics completely takes care of this in the background, automatically detecting if a metric is being updated on multiple devices and accumulating the states from different devices before reporting the calculated metric to the user.
```

Another core feature of TorchMetrics is its ability to scale to multiple devices seamlessly. Modern deep learning models are often trained on hundreds of devices GPUs or TPUs (see [@large_example1; @large_example2] for examples). This scale introduces the need to synchronize metrics across machines to get the correct value during training and evaluation. In distributed environments, TorchMetrics automatically accumulates across devices before reporting the calculated metric to the user.

**Review by williamFalcon — APPROVED** [review]

**Comment by williamFalcon** [comment]
nice job everyone!

**Pull request merged**